src/session.c

```diff
@@ -0,0 +1,575 @@
+                                   CK_SESSION_HANDLE *session)
+{
+    CK_FUNCTION_LIST *f;
+    bool wait_ok = true;
+    CK_ULONG cs = 0;
+    uint64_t startime = 0;
+    CK_RV ret;
+
+    ret = p11prov_ctx_status(p->provctx, &f);
+    if (ret != CKR_OK) {
+        return ret;
+    }
+
+    while (wait_ok) {
+        if (cs == 0) {
+            cs = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+            if (cs > p->max_sessions) {
+                P11PROV_debug_once("Max Session (%lu) exceeded!", cs);
+                (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+                ret = CKR_SESSION_COUNT;
+                cs = 0;
+                wait_ok = cyclewait_with_timeout(MAX_WAIT, SLEEP, &startime);
+                continue;
+            }
+        }
+
+        wait_ok = false;
+        ret = f->C_OpenSession(slot, CKF_SERIAL_SESSION, NULL, NULL, session);
+        P11PROV_debug("C_OpenSession ret:%lu (session: %lu)", ret, session);
+        if (ret == CKR_SESSION_COUNT) {
+            wait_ok = cyclewait_with_timeout(MAX_WAIT, SLEEP, &startime);
+        }
+    }
+
+    if (ret != CKR_OK && cs != 0) {
+        (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+    }
+    return ret;
+}
+
+static void internal_session_close(P11PROV_SESSION *session)
+{
+    if (session->session != CK_INVALID_HANDLE) {
+        CK_FUNCTION_LIST *f;
+        CK_RV ret;
+
+        ret = p11prov_ctx_status(session->provctx, &f);
+        if (ret == CKR_OK) {
+            ret = f->C_CloseSession(session->session);
+            if (ret != CKR_OK) {
+                P11PROV_raise(session->provctx, ret,
+                              "Failed to close session %lu", session->session);
+            }
+        }
+        /* regardless of the result the sesssion is gone */
+        if (session->pool) {
+            (void)__atomic_fetch_sub(&session->pool->cur_sessions, 1,
+                                     __ATOMIC_SEQ_CST);
+        }
+        session->session = CK_INVALID_HANDLE;
+    }
+}
+
+CK_RV p11prov_session_pool_init(P11PROV_CTX *ctx, CK_TOKEN_INFO *token,
+                                P11PROV_SESSION_POOL **_pool)
+{
+    P11PROV_SESSION_POOL *pool;
+
+    P11PROV_debug("Creating new session pool");
+
+    pool = OPENSSL_zalloc(sizeof(P11PROV_SESSION_POOL));
+    if (!pool) {
+        return CKR_HOST_MEMORY;
+    }
+    pool->provctx = ctx;
+
+    pthread_mutex_init(&pool->lock, 0);
+
+    if (token->ulMaxSessionCount != CK_EFFECTIVELY_INFINITE
+        && token->ulMaxSessionCount != CK_UNAVAILABLE_INFORMATION) {
+        pool->max_sessions = token->ulMaxSessionCount;
+        /* keep a max of 10% of the sessions */
+        pool->limit_sessions = pool->max_sessions / 10;
+        if (pool->limit_sessions == 0) {
+            pool->limit_sessions = 3;
+        }
+    } else {
+        /* arbitrary max concurrent open sessions */
+        pool->max_sessions = 1024;
+        pool->limit_sessions = 8;
```

Comment:
I am not sure why we want to tie that to the number of CPUs?
TBH, I pulled this values more or less at random, I think 1024 is way too high as a default maximum, but 8 sounded like a decent number of concurrent sessions to have, enough to have a login session, a couple of keys and perform 2,3 operations in parallel and still have a session to spare.
If we want to tie this to the CPU I would think we'd have to double the number of CPUs + 1 (login session), but then again I am not sure why we'd do that, I have no data to know how many concurrent session a single application would normally use.

I thought that in future we could keep some statistics and self tune.
This code *needs* to be improved because on a test run I did, reading the logs, I discovered we are not using the session objects with an actually cached session in preference, instead we just pick the first free and opena  new session if one is not available.

So this code definitely needs more smarts when it is looking for a free session.